lib/tree.rb

```diff
@@ -16,42 +16,132 @@ def initialize
+  # Time Complexity: O(log n)
+  # Space Complexity: [O(1)-if it's a loop] - why do I wanna say O(n)?? help me. 
+  # Try with a while loop
+  def add(key, value = nil)
+    if @root.nil?
+      @root = TreeNode.new(key, value)
+    else
+      add_helper(@root, key, value)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, key, value)
+    # if current = return TreeNode.new
+    if key < current.key
+      # if current.left is nil place the new node
+      # otherwise make current.left the new current and run through the same checks
+      if current.left.nil? 
+        current.left = TreeNode.new(key, value)
+        return
+      else
+        add_helper(current.left, key, value)
+      end
+    else 
+      if current.right.nil? 
+        current.right = TreeNode.new(key, value)
+        return
+      else
+        add_helper(current.right, key, value)
+      end
+    end
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current, key)
+    if current.nil?
+      return nil
+    end
+    if current.key == key
+      return current.value
+    end
+    if key > current.key 
+      return find_helper(current.right, key)
+    else
+      return find_helper(current.left, key)
+    end 
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+    # left-root-right
   def inorder
-    raise NotImplementedError
+    # current = @root
+    inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, values) # Values is an array
+    if current.nil?
+      return values
+    end
+    inorder_helper(current.left, values)
+    values.push({:key=>current.key, :value=>current.value})
+    inorder_helper(current.right, values)
+    return values
+  end
+
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+    # root-left-right
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current, values)
+    if current.nil?
+      return values
+    end
+    values.push({:key=>current.key, :value=>current.value})
+    preorder_helper(current.left, values)
+    preorder_helper(current.right, values)
+    return values
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+    # left-right-root
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current, values)
+    if current.nil?
+      return values
+    end
+    postorder_helper(current.left, values)
+    postorder_helper(current.right, values)
+    values.push({:key=>current.key, :value=>current.value})
+    return values
+  end
+
+  # Time Complexity: O(n)^2
+  # Space Complexity: O(n)
   def height
```

Comment:
👍 The time complexity is O(n) (not n^2)

The space complexity is O(log n) if the tree is balanced or O(n) if it's not.  The space complexity is basically the height of the tree.